alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt

```diff
@@ -242,121 +205,131 @@ class DocumentLevelMonitorRunner : MonitorRunner() {
-                    inputRunResults,
-                    docsToQueries,
-                )
-            }
-            monitorResult = monitorResult.copy(inputResults = InputRunResults(listOf(inputRunResults)))
-
-            /*
-             populate the map queryToDocIds with pairs of <DocLevelQuery object from queries in monitor metadata &
-             list of matched docId from inputRunResults>
-             this fixes the issue of passing id, name, tags fields of DocLevelQuery object correctly to TriggerExpressionParser
-             */
-            queries.forEach {
-                if (inputRunResults.containsKey(it.id)) {
-                    queryToDocIds[it] = inputRunResults[it.id]!!
-                }
-            }
-
-            val idQueryMap: Map<String, DocLevelQuery> = queries.associateBy { it.id }
+                    val shards = mutableSetOf<String>()
+                    shards.addAll(indexUpdatedRunContext.keys)
+                    shards.remove("index")
+                    shards.remove("shards_count")
 
-            val triggerResults = mutableMapOf<String, DocumentLevelTriggerRunResult>()
-            // If there are no triggers defined, we still want to generate findings
-            if (monitor.triggers.isEmpty()) {
-                if (dryrun == false && monitor.id != Monitor.NO_ID) {
-                    createFindings(monitor, monitorCtx, docsToQueries, idQueryMap, true)
-                }
-            } else {
-                monitor.triggers.forEach {
-                    triggerResults[it.id] = runForEachDocTrigger(
+                    val nodeMap = getNodes(monitorCtx)
+                    val nodeShardAssignments = distributeShards(
                         monitorCtx,
-                        monitorResult,
-                        it as DocumentLevelTrigger,
-                        monitor,
-                        idQueryMap,
-                        docsToQueries,
-                        queryToDocIds,
-                        dryrun,
-                        executionId = executionId,
-                        workflowRunContext = workflowRunContext
+                        nodeMap.keys.toList(),
+                        shards.toList(),
+                        concreteIndexName
                     )
+
+                    val responses: Collection<DocLevelMonitorFanOutResponse> = suspendCoroutine { cont ->
+                        val listener = GroupedActionListener(
+                            object : ActionListener<Collection<DocLevelMonitorFanOutResponse>> {
+                                override fun onResponse(response: Collection<DocLevelMonitorFanOutResponse>) {
+                                    cont.resume(response)
+                                }
+
+                                override fun onFailure(e: Exception) {
+                                    if (e.cause is Exception)
+                                        cont.resumeWithException(e.cause as Exception)
```

Comment:
can't do resume with exception here i guess. 
Grouped action listener doesn't tolerate partial failure. 
we might need to implement a new grouped listener which will return a list of responses and list of exceptions rather than one failure nullifying rest of the responses.

Reply:
this is handled now. https://github.com/opensearch-project/alerting/blob/7356f94d33cd6af97a99a4bad656b52d0911909f/alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt#L310
https://github.com/opensearch-project/alerting/blob/7356f94d33cd6af97a99a4bad656b52d0911909f/alerting/src/main/kotlin/org/opensearch/alerting/DocumentLevelMonitorRunner.kt#L353-#L356